### DISCUSSION TITLE
NPC Dialog

### DISCUSSION THREAD

**Discussion by Zintixx** [AlanMorel/MapleServer2#345] [General]
Let's talk NPC Dialog.
I've come to the conclusion that the _initial_ dialog is handled via the server and the majority is handled by what we currently have (content count, distractors and gotos.

First let's talk what's currently missing:

- script `randomPick` attribute
	- What is randomPick for? It doesn't seem to be the "initial" dialogue often times.
- content `functionID` attribute
	- I had dived into this before. If an NPC has a "functionID" in their dialog, this means that a specific function needs to happen for this particular NPC (ex. Kay's Wheel or Natalie the Archaeologist both have functionIDs). If there's more than functionID, it'll be handled by each respective ID.
- content `buttonSet` attribute
	- This is pretty straight forward. If there is a `buttonSet` attribute present, this supersedes  the normally functionality of setting the next button.
	- Go to Fails. This is not implemented at the moment but with how every goto fail is specific to that certain dialog (if Player is missing a certain item, player doesn't have a quest, etc other requirements), this seems like it needs to be handled server side. 

#### Example 1: NPC `Ropey`
This NPC is located in Tria (top left). This NPC has a shop.
![enter image description here](https://i.imgur.com/8Eu29iA.png)
This is normally how Ropey's dialog should be. An option for opening his shop and to talk
![enter image description here](https://i.imgur.com/n9UV4cc.png)
Here's the dialog if you press Talk.
Now let's look at his script (just a snip)
```	
	<select id="0" feature="" locale="">
		<content text="$script:1016142906000509$" />
	</select>
	<script randomPick="1" id="30" feature="" locale="">
		<content text="$script:1016142906000512$">
			<distractor text="$script:1016142906000513$" goto="31" gotoFail="" />
			<distractor text="$script:1016142906000514$" goto="34" gotoFail="" />
		</content>
	</script>
```
Ropey's first dialog is in id 0.
And if you talk with him, you'll get id 30 dialog.
Let's look at `Surnuny` in Tria, another shop NPC.
Surnuny doesn't have any dialog, but if we take a look at his dialog script, this is all we see:
```	
	<select id="0" feature="" locale="">
		<content text="$script:0831180306000001$" />
	</select>
```

Perhaps if NPC shopID != 0 & randomPick doesn't exist, it should just pop up the shop? Else, have an option similar to `Ropey`? Maybe.

#### Example 2:  `Orde` In Grand Hall
Order in grand hall has different dialogs depending on what job you are.
If Striker:
![enter image description here](https://i.imgur.com/bEkBfjX.png)

If Assassin:
![enter image description here](https://i.imgur.com/qowxwSU.png)

(Not posting the script for this one. It's pretty long. NPC ID: 11000042)
If you take a look at the script for this NPC, you'll see there's _multiple_ randomPicks of dialogs with no real indication on how certain classes get certain dialogs.

Not to mention, Nam also posted about this in this issue: https://github.com/AlanMorel/MapleServer2/issues/313

Thoughts?

**Comment by gwagwaflamingo** [comment]
> What is randomPick for? It doesn't seem to be the "initial" dialogue often times.

Frankly, this fries my brain too. No idea what is the deal with it. But according to O3O3O...
![image](https://user-images.githubusercontent.com/72739828/130952394-56a9311e-8b8e-48e8-a904-193f93f40fb6.png)

> Perhaps if NPC shopID != 0 & randomPick doesn't exist, it should just pop up the shop?

Not too sure what you mean with this, could you rephrase it?
Select are simply dialog that show up when the player has more than one option to pick from - such as quests, shops and scripted dialogs - and almost every NPC has one; with the only exceptions being NPCs that never are assigned as quest assigners/progressers/receivers, don't have shops. or don't have dialogue (an example being 11000133, a NPC script that has only monologues).

> Order in grand hall has different dialogs depending on what job you are.

Orde is a **function npc**, which _apparently_ are always hardcoded.    real ms1 moment

**Reply by Zintixx** [reply]
> Not too sure what you mean with this, could you rephrase it?
> Select are simply dialog that show up when the player has more than one option to pick from - such as quests, shops and scripted dialogs - and almost every NPC has one; with the only exceptions being NPCs that never are assigned as quest assigners/progressers/receivers, don't have shops. or don't have dialogue (an example being 11000133, a NPC script that has only monologues).

It's all theoretical but I always thought `randomPick` is the first dialog that should be presented. Assuming this is true, this would make sense for the NPC shop like Surnuny in Tria (Because he doesn't have a dialog option with a `randomPick` attribute).

> > Order in grand hall has different dialogs depending on what job you are.
> 
> Orde is a **function npc**, which _apparently_ are always hardcoded. real ms1 moment

So what makes this NPC a function npc? Is there any indication of it on the xmls?

**Reply by gwagwaflamingo** [reply]
> So what makes this NPC a function npc? Is there any indication of it on the xmls?

All of her dialog strings are in scriptnpcfunction.xml, which is why I assumed she is one.

**Comment by AngeloTadeucci** [comment]
Any updates after PR #385 got merged?
